### ISSUE TITLE
GitLens breaks Outline for Python

### ISSUE THREAD

**Issue by rjbruin** [eamodio/vscode-gitlens#439]
- GitLens Version: 8.4.1
- VSCode Version: 1.24.1
- OS Version: Ubuntu 16.04

Steps to Reproduce:

1. Install Python and GitLens extensions
2. Open Outline view
3. Open Python file in editor. This file should be displayed correctly in Outline.
4. Open another Python file. Now the Outline stays blank.
5. Switching to Search or Debug, then back to Explorer returns the Outline for the open file, but switching to another Python file once again breaks Outline.

I have checked all extensions I'm using, and this occurs only when Python and GitLens are enabled together. The problem does not occur when disabling GitLens.

**Comment by eamodio** [comment]
I'm pretty sure this is yet another case of https://github.com/DonJayamanne/pythonVSCode/issues/553 which was never fixed as far as I know.

**Comment by eamodio** [comment]
Opened this on the MS repo: https://github.com/Microsoft/vscode-python/issues/2112

**Comment by eamodio** [comment]
@rjbruin I believe Microsoft released a completely new version of the Python language server (as part of the Python extension) yesterday or today. Can you try it and see if this is still an issue?

**Comment by rjbruin** [comment]
Unfortunately the problem still persists with `vscode-python 2018.7.0` and `GitLens 8.5.2`.

**Comment by eamodio** [comment]
Damn :(

**Comment by ericsnowcurrently** [comment]
@rjbruin, make sure you have `"python.jediEnabled": false` in your settings.  That will activate the new language server.

**Comment by rjbruin** [comment]
@ericsnowcurrently Thanks! 

It seems the issue is connected to the new language server:

- No GitLens + `"python.jediEnabled": true`: works;
- No GitLens + `"python.jediEnabled": false`: no file outline will load anymore, and switching out of Explorer and back doesn't return the outline anymore

However, even when disabling the new language server and enabling GitLens the issue remains. Even more confusing, I'm not seeing the same behavior anymore even with the settings of the original issue:

- GitLens + `"python.jediEnabled": true` (original issue): certain files load without problem, while others don't. Things I've tried to narrow down the differences:
  - Both files are in a repository tracked by GitLens: the one that breaks outline is in our main company repository, the one that works is in a new local repository;
  - Both files have the exact same content and name;
  - Both workspace folders use the same Python interpreter (from virtualenv)

I'm getting kind of lost here. As for the GitLens issue, it seems to have to do with the repository the file is in. If I can do anything to help debug this, please let me know!

**Comment by ericsnowcurrently** [comment]
@rjbruin, if it's okay, let's discuss this on Microsoft/vscode-python#2112.  If you want to keep the discussion here I can work with that. :)

**Comment by rjbruin** [comment]
@ericsnowcurrently I'm fine with that, though I assume you only want to talk about the language server there and leave the GitLens issue here?

**Comment by ericsnowcurrently** [comment]
Correct.  I just don't want to clutter up this issue with stuff that doesn't relate to GitLens. :)

**Comment by prosenboim** [comment]
All related issues are marked as closed, but problem is still there with GitLens 9.7.3, Python 2019.4.12954 and VS code Version: 1.34.20-insider

**Comment by astrowonk** [comment]
> All related issues are marked as closed, but problem is still there with GitLens 9.7.3, Python 2019.4.12954 and VS code Version: 1.34.20-insider

Yeah, I'm being forced to use Jedi because of [another bug](https://github.com/Microsoft/vscode-python/issues/2528) in Python Language Server, and so have had to turn off GitLens. :-(

**Comment by DonJayamanne** [comment]
> Yeah, I'm being forced to use Jedi because of another bug in Python Language Server, and so have had to turn off GitLens. :-(

@eamodio When using the `Python Language Server`, there's no extension code involved. Its purely Language Server (protocol that communicates with VS Code). So I don't see how this would be an extension issue. 
I.e. it seem this issue occurs whether users use language server or not (in Python Extension - i.e. it isn't any typescript code in Python Extension).

I'd say this is an issue upstream on VS Code.

**Comment by eamodio** [comment]
@DonJayamanne Last I looked into this (which was a LONG time ago), the core issue was that multiple places (GitLens, outline view, etc) call into the language server (via executeDocumentSymbolProvider) which only returned valid symbols for one request. So sometimes that would be GitLens, and the outline view would break, or sometimes it was the outline view, then GitLens code lens wouldn't show. And it seemed to only really manifest itself if the calls were in close succession -- like it was actually processing one request while the other one came in.

**Comment by Daniel451** [comment]
@eamodio yeah, can confirm this. Guess this should be fixed in the language server API?! Is there an issue related to this?

**Comment by DonJayamanne** [comment]
>  which only returned valid symbols for one request. So sometimes that would be GitLens, and the outline view would break, or sometimes it was the outline view, then GitLens code lens wouldn't show. And it seemed to only really manifest itself if the calls were in close succession -- like it was actually processing one request while the other one came in.

Wouldn't this be an issue with VS Code. After all, you're invoking VS Code API.

**Comment by eamodio** [comment]
AFAIK, the vscode api just calls into the extension/language servers (those registered with `registerDocumentSymbolProvider`) so if no data is returned by them, there is nothing vscode can do.

**Comment by eamodio** [comment]
@DonJayamanne could this be the cause:

https://github.com/microsoft/vscode-python/blob/3914055173bb7823b1db0ac17251b871764112b9/src/client/providers/symbolProvider.ts#L89-L94

It looks like if 2 requests come in quickly the first one will be resolved with an empty array.

**Comment by awschristou** [comment]
> It looks like if 2 requests come in quickly the first one will be resolved with an empty array.
>
> Not sure why both requests wouldn't just resolve to the same promise.

I wonder if the subsequent request is expecting that the file has been modified since the first request

**Comment by DonJayamanne** [comment]
> It looks like if 2 requests come in quickly the first one will be resolved with an empty array.

Agreed, will fix that.

**Comment by eamodio** [comment]
@DonJayamanne 
The issue it causes, is that if GitLens and say the Outline view make a call to get the document symbols in quick succession one of them will get results and the other will get an empty array.

**Comment by yakovyashfe** [comment]
I changed Language server to jedi, now I have an outline
